File: zgc.py
"""ZGC coloured-pointer decoder.

Two pointer formats ship across the JDK matrix:

* **Classic ZGC** (JDK 15..20, JDK 17 here) — colour bits live in HIGH
  positions. The address is the low 42 bits:
      heap_addr = raw & ZAddressOffsetMask
  Live pointers satisfy `raw & ZAddressGoodMask != 0`.

* **Generational ZGC** (JDK 21+ with `-XX:+ZGenerational`, JDK 25 default) —
  colour bits live in LOW positions. Uncoloring is a right-shift:
      heap_addr = raw >> ZPointerLoadShift
  Live pointers satisfy `raw & ZPointerLoadBadMask == 0`.

The per-pointer shift-table variant HotSpot uses inside JIT-compiled code
isn't needed from outside: we always work against load-good pointers, so
a single global shift/mask works.

Activation is detected by checking `ZGlobalsForVMStructs::_instance_p`
and which mask fields are populated. With JDK 21 `-XX:+UseZGC` but no
`-XX:+ZGenerational`, the struct exists but fields stay zero — in that
legacy mode we report ZGC-as-unsupported and bail.
"""
from __future__ import annotations

from dataclasses import dataclass

from vm_meta import VMMeta, _ptr, _u64


@dataclass
class ZGCDecoder:
    vm: VMMeta
    mode: str              # "classic" | "generational" | "off"
    instance: int          # ZGlobalsForVMStructs* (0 if off)
    offset_mask_ptr: int   # ptr-to-uintptr_t for ZAddressOffsetMask (classic)
    load_shift_ptr: int    # ptr-to-size_t for ZPointerLoadShift (generational)
    load_bad_mask_ptr: int  # optional, for liveness check
    store_good_ptr: int = 0  # generational: ZPointerStoreGoodMask
    address_good_ptr: int = 0  # classic: ZAddressGoodMask (for re-color)

    @classmethod
    def detect(cls, vm: VMMeta) -> ZGCDecoder:
        if not vm.has_type("ZGlobalsForVMStructs"):
            return cls(vm, "off", 0, 0, 0, 0)
        zg = vm.type("ZGlobalsForVMStructs")
        inst = _ptr(vm.reader, zg.static_field("_instance_p").address)
        if not inst:
            return cls(vm, "off", 0, 0, 0, 0)

        # Probe which variant has meaningful values set.
        def ptr_of(field: str) -> int:
            if not zg.has_field(field):
                return 0
            return _ptr(vm.reader, inst + zg.field(field).offset)

        gen_shift_ptr = ptr_of("_ZPointerLoadShift")
        gen_good_ptr = ptr_of("_ZPointerLoadGoodMask")
        gen_bad_ptr = ptr_of("_ZPointerLoadBadMask")
        classic_offset_ptr = ptr_of("_ZAddressOffsetMask")
        classic_good_ptr = ptr_of("_ZAddressGoodMask")
        classic_bad_ptr = ptr_of("_ZAddressBadMask")

        # Store-side masks for generational (used when we need to encode
        # a wide heap address back into a coloured pointer for writing).
        store_good_ptr = ptr_of("_ZPointerStoreGoodMask")

        # Generational: LoadGoodMask populated (non-zero) and LoadShift > 0.
        if gen_good_ptr and _u64(vm.reader, gen_good_ptr) != 0 \
                and gen_shift_ptr and _u64(vm.reader, gen_shift_ptr) > 0:
            return cls(vm, "generational", inst,
                       offset_mask_ptr=0,
                       load_shift_ptr=gen_shift_ptr,
                       load_bad_mask_ptr=gen_bad_ptr,
                       store_good_ptr=store_good_ptr)

        # Classic: AddressGoodMask populated AND the field is actually present.
        if classic_good_ptr and _u64(vm.reader, classic_good_ptr) != 0 \
                and classic_offset_ptr:
            return cls(vm, "classic", inst,
                       offset_mask_ptr=classic_offset_ptr,
                       load_shift_ptr=0,
                       load_bad_mask_ptr=classic_bad_ptr,
                       address_good_ptr=classic_good_ptr)

        # ZGlobals exists but both variants are zero — legacy ZGC with no
        # populated masks (JDK 21 `-XX:+UseZGC` only). Can't decode safely.
        return cls(vm, "off", inst, 0, 0, 0)

    @property
    def is_active(self) -> bool:
        return self.mode != "off"

    def decode(self, raw: int) -> int:
        """Return a virtual address that can be fed to ReadProcessMemory.

        * Generational: colour bits sit in the low `LoadShift` bits; a right
          shift by the current shift value yields the heap offset, and the
          OS-mapped heap lives at that value (no base is added).

        * Classic: the colour bits are already in high positions, and ZGC
          maps the heap virtual-three-times (Marked0/Marked1/Remapped). A
          load-good pointer is directly dereferenceable — we just pass it
          through. Masking off the high good-bit would de-map it.

        Re-reads the shift on every call so that a GC phase flip between
        calls doesn't leave us shifting by a stale value. If the pointer is
        load-bad (flip happened *during* our read), retries once with the
        fresh shift — good enough for 99.99% of cases; pathological flip
        races require the per-pointer cpu_shift_table (M11 backlog).
        """
        if raw == 0 or self.mode == "off":
            return raw
        if self.mode == "classic":
            return raw
        # generational path with a one-shot flip retry.
        for _ in range(2):
            shift = _u64(self.vm.reader, self.load_shift_ptr)
            candidate = raw >> shift
            if not self._is_load_bad_now(raw):
                return candidate
        return raw >> _u64(self.vm.reader, self.load_shift_ptr)

    def _is_load_bad_now(self, raw: int) -> bool:
        if not self.load_bad_mask_ptr:
            return False
        return (raw & _u64(self.vm.reader, self.load_bad_mask_ptr)) != 0

    def encode_for_store(self, heap_addr: int) -> int:
        """Re-colour a wide heap address into the format a ZGC store-site
        would use. For generational ZGC: shift into the address range
        and OR in the store-good colour bits. For classic: OR in the
        current good mask.

        Result is the raw value to write into the target's reference slot.
        """
        if heap_addr == 0 or self.mode == "off":
            return heap_addr
        if self.mode == "generational":
            if not self.store_good_ptr:
                # Fallback: reuse load-good if store-good wasn't exported.
                store_good_ptr = self.store_good_ptr
            else:
                store_good_ptr = self.store_good_ptr
            store_good = _u64(self.vm.reader, store_good_ptr) if store_good_ptr else 0
            shift = _u64(self.vm.reader, self.load_shift_ptr)
            return ((heap_addr << shift) | store_good) & 0xFFFFFFFFFFFFFFFF
        # classic: OR the current AddressGoodMask onto the offset.
        if self.address_good_ptr:
            good = _u64(self.vm.reader, self.address_good_ptr)
            offset_mask = (_u64(self.vm.reader, self.offset_mask_ptr)
                           if self.offset_mask_ptr else 0xFFFFFFFFFFFFFFFF)
            return (heap_addr & offset_mask) | good
        return heap_addr

    def is_load_bad(self, raw: int) -> bool:
        """True if the pointer carries a "bad" colour and would need a
        barrier before use. We usually only see load-good values."""
        if self.mode != "generational" or not self.load_bad_mask_ptr:
            return False
        bad_mask = _u64(self.vm.reader, self.load_bad_mask_ptr)
        return (raw & bad_mask) != 0

    def __repr__(self) -> str:
        return f"ZGCDecoder(mode={self.mode!r}, instance={self.instance:#x})"

File: cpp/src/jvm/zgc.cpp
#include "zgc.hpp"

namespace marrow {

ZGCDecoder::ZGCDecoder(VMMeta* vm, Mode mode, uint64_t instance,
                       uint64_t offset_mask_ptr, uint64_t load_shift_ptr,
                       uint64_t load_bad_mask_ptr,
                       uint64_t store_good_ptr, uint64_t address_good_ptr)
    : vm_(vm), mode_(mode), instance_(instance),
      offset_mask_ptr_(offset_mask_ptr),
      load_shift_ptr_(load_shift_ptr),
      load_bad_mask_ptr_(load_bad_mask_ptr),
      store_good_ptr_(store_good_ptr),
      address_good_ptr_(address_good_ptr) {}

ZGCDecoder ZGCDecoder::detect(VMMeta* vm)
{
    if (!vm->has_type("ZGlobalsForVMStructs")) return ZGCDecoder{};
    const TypeInfo* zg = vm->type("ZGlobalsForVMStructs");
    if (!zg->has_field("_instance_p")) return ZGCDecoder{};
    uint64_t inst = vm->reader()->read_u64(zg->field("_instance_p")->address);
    if (!inst) return ZGCDecoder{};

    auto ptr_of = [&](const char* name) -> uint64_t {
        if (!zg->has_field(name)) return 0;
        return vm->reader()->read_u64(inst + zg->field(name)->offset);
    };

    uint64_t gen_shift   = ptr_of("_ZPointerLoadShift");
    uint64_t gen_good    = ptr_of("_ZPointerLoadGoodMask");
    uint64_t gen_bad     = ptr_of("_ZPointerLoadBadMask");
    uint64_t cls_offset  = ptr_of("_ZAddressOffsetMask");
    uint64_t cls_good    = ptr_of("_ZAddressGoodMask");
    uint64_t cls_bad     = ptr_of("_ZAddressBadMask");
    uint64_t gen_store   = ptr_of("_ZPointerStoreGoodMask");

    if (gen_good && vm->reader()->read_u64(gen_good) != 0
            && gen_shift && vm->reader()->read_u64(gen_shift) > 0) {
        return ZGCDecoder(vm, Mode::Generational, inst,
                          /*offset*/0, gen_shift, gen_bad, gen_store, 0);
    }
    if (cls_good && vm->reader()->read_u64(cls_good) != 0 && cls_offset) {
        return ZGCDecoder(vm, Mode::Classic, inst,
                          cls_offset, 0, cls_bad, 0, cls_good);
    }
    return ZGCDecoder{};
}

bool ZGCDecoder::is_load_bad_now(uint64_t raw)
{
    if (!load_bad_mask_ptr_) return false;
    return (raw & vm_->reader()->read_u64(load_bad_mask_ptr_)) != 0;
}

uint64_t ZGCDecoder::decode(uint64_t raw)
{
    if (raw == 0 || mode_ == Mode::Off) return raw;
    if (mode_ == Mode::Classic) return raw; // already dereferenceable
    for (int retry = 0; retry < 2; ++retry) {
        uint64_t shift = vm_->reader()->read_u64(load_shift_ptr_);
        uint64_t cand = raw >> shift;
        if (!is_load_bad_now(raw)) return cand;
    }
    return raw >> vm_->reader()->read_u64(load_shift_ptr_);
}

uint64_t ZGCDecoder::encode_for_store(uint64_t heap_addr)
{
    if (heap_addr == 0 || mode_ == Mode::Off) return heap_addr;
    if (mode_ == Mode::Generational) {
        uint64_t store_good = store_good_ptr_ ? vm_->reader()->read_u64(store_good_ptr_) : 0;
        uint64_t shift = vm_->reader()->read_u64(load_shift_ptr_);
        return (heap_addr << shift) | store_good;
    }
    if (address_good_ptr_) {
        uint64_t good = vm_->reader()->read_u64(address_good_ptr_);
        uint64_t mask = offset_mask_ptr_
            ? vm_->reader()->read_u64(offset_mask_ptr_)
            : 0xFFFFFFFFFFFFFFFFull;
        return (heap_addr & mask) | good;
    }
    return heap_addr;
}

bool ZGCDecoder::is_load_bad(uint64_t raw)
{
    if (mode_ != Mode::Generational || !load_bad_mask_ptr_) return false;
    uint64_t bad = vm_->reader()->read_u64(load_bad_mask_ptr_);
    return (raw & bad) != 0;
}

} // namespace marrow
